rmw_zenoh_cpp/src/rmw_zenoh.cpp

```diff
@@ -1630,145 +1662,695 @@ rmw_return_loaned_message_from_subscription(
+      max_data_length,
+      allocator->state));
+  if (!request_bytes) {
+    RMW_SET_ERROR_MSG("failed allocate request message bytes");
+    allocator->deallocate(request_bytes, allocator->state);
+    return RMW_RET_ERROR;
+  }
+
+  // Object that manages the raw buffer
+  eprosima::fastcdr::FastBuffer fastbuffer(request_bytes, max_data_length);
+
+  // Object that serializes the data
+  eprosima::fastcdr::Cdr ser(
+    fastbuffer,
+    eprosima::fastcdr::Cdr::DEFAULT_ENDIAN,
+    eprosima::fastcdr::Cdr::DDS_CDR);
+  if (!client_data->request_type_support->serialize_ros_message(
+      ros_request,
+      ser,
+      client_data->request_type_support_impl))
+  {
+    allocator->deallocate(request_bytes, allocator->state);
+    return RMW_RET_ERROR;
+  }
+
+  size_t data_length = ser.getSerializedDataLength();
+
+  // TODO(francocipollone): Do I really need the sequency number here?
+  *sequence_id = 0;
+
+  // Send request
+  z_get_options_t opts = z_get_options_default();
+  opts.target = Z_QUERY_TARGET_ALL;
+  opts.value.payload = z_bytes_t{data_length, reinterpret_cast<const uint8_t *>(request_bytes)};
+  opts.value.encoding = z_encoding(Z_ENCODING_PREFIX_EMPTY, NULL);
+  client_data->zn_closure_reply = z_closure(client_data_handler, nullptr, client_data);
+  z_get(
+    z_loan(context_impl->session), z_loan(
+      client_data->keyexpr), "", &client_data->zn_closure_reply, &opts);
+
+  return RMW_RET_OK;
+}
+
+//==============================================================================
+/// Take an incoming ROS service response.
+rmw_ret_t
+rmw_take_response(
+  const rmw_client_t * client,
+  rmw_service_info_t * request_header,
+  void * ros_response,
+  bool * taken)
+{
+  *taken = false;
+  RCUTILS_LOG_INFO_NAMED("rmw_zenoh_cpp", "[rmw_take_response]");
+
+  RMW_CHECK_ARGUMENT_FOR_NULL(client, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(client->data, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(ros_response, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(taken, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    client,
+    client->implementation_identifier,
+    rmw_zenoh_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    client->service_name, "client has no service name", RMW_RET_INVALID_ARGUMENT);
+
+  rmw_client_data_t * client_data = static_cast<rmw_client_data_t *>(client->data);
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    client->data, "Unable to retrieve client_data from client.", RMW_RET_INVALID_ARGUMENT);
+
+  z_owned_reply_t * latest_reply = nullptr;
+
+  std::lock_guard<std::mutex> lock(client_data->message_mutex);
+  if (client_data->replies.empty()) {
+    // TODO(francocipollone): Verify behavior.
+    RCUTILS_LOG_ERROR_NAMED("rmw_zenoh_cpp", "[rmw_take_response] Response message is empty");
+    return RMW_RET_ERROR;
+  }
+  latest_reply = &client_data->replies.back();
```

Comment:
This is always responding to the latest_reply.  Are we sure that that is always going to be for *this* client?  That is, if we had multiple clients all sending requests at the same time, how do we know that we've taken the correct one?

Reply:
> if we had multiple clients all sending requests at the same time, how do we know that we've taken the correct one?

This is solved at the server side, as we are using queryables, from the server side we store the query and answer to it using the same channel.

This latest reply and the vector replies are here to support the situation when you have multiple servers with the same keyexpr. In this part you might get multiple replies (one from each server). We are not tackling this situation in this PR, we simply return the latest reply.